lib/ide/parser/src/api.rs

```diff
@@ -3,9 +3,53 @@
 use crate::prelude::*;
 
 use ast::IdMap;
+use ast::HasRepr;
+use ast::HasIdMap;
 
 pub use ast::Ast;
 
+use serde::Deserialize;
+use serde::Serialize;
+
+
+// ============
+// == Module ==
+// ============
+
+/// Parsed file / module with metadata
+#[derive(Debug,Clone,Serialize,Deserialize)]
+pub struct ModuleWithMetadata {
+    /// ast representation
+    pub ast: Ast,
+    /// raw metadata in json
+    pub metadata: serde_json::Value
+}
+
+const ID_TAG       : &str = "# [idmap] ";
+const METADATA_TAG : &str = "# [metadata] ";
+
+impl ToString for ModuleWithMetadata {
+    fn to_string(&self) -> String {
+        let remove_newlines = |string:String| string
+            .chars()
+            .filter(|c| c != &'\n' && c != &'\r')
+            .collect::<String>();
+
+        let code = self.ast.repr();
+        let ids  = remove_newlines(
+            serde_json::to_string(&self.ast.id_map()).expect(
+                "It should be possible to serialize idmap."
+            )
```

Comment:
when it can fail? I do not like panicking in the code. If this could really happen (like for example file with bad text inside), this should result in a data-type error and this erro should be handled - definitely not panicking.

Reply:
What kind of bad text?